### ISSUE TITLE
Expand rewardable object functionality

### ISSUE THREAD

**Issue by Nordsoft91** [vcmi/vcmi#1289]
Currently it's possible to configure specific rewards. It's needed to be extended with:
- [x] random skill/spell/creature from list/level/school
- [x] random skill/spell/creature except from list/level/school
- [x] movement points next day
- [x] executing advmap spell (town gate for example)
- [x] improving random skill (limited by advanced/expert)
- [ ] reduce movement penalty (specific terrain) (day(s)/eow)
- [x] reduce movement penalty (any terrain) (day(s)/eow)
- [x] mana points
- [x] scouting radius (day(s)/eow)
- [x] scouting radius (instant, observatory-like)
- [x] hiding terrain (Cover of Darkness)
- [ ] creatures recruitment (dwelling-like)
- [ ] limiter for number of empty slots in army

**Comment by misiokles** [comment]
I would also to see negative bonuses after visiting some objects, like -luck for Crypt-like buildings.

**Comment by IvanSavenko** [comment]
Needs rechecking, probably half of these options are already possible

**Comment by rilian-la-te** [comment]
@misiokles it is already possible. 

@Nordsoft91 to movement points - just set MOVEMENT bonus for one day duration with additive value.
For reduce movement penalty - also use scheme like this with bonuses. But now ROUGH_TERRAIN_DISCOUNT only used to all terrains.

@IvanSavenko makes sense to add ROUGH_TERRAIN_DISCOUNT a terrain subtype for such task like " reduce movement penalty (specific terrain/all) (day(s)/eow)"

Mana points is possible as is.

Scouting radius - use SIGHT_RADIUS bonus, also seems possible.

**Comment by DjWarmonger** [comment]
Yep, scouting radius is possible (as any other bonus, especially after secondary skill rework).

However, **Ruins** from HotA actually reveal terrain within a radius immediately and has nothing to do with hero scouting.

**Comment by IvanSavenko** [comment]
Updated task.
Out of remaining ones these do sound interesting:
- reduce movement penalty (specific terrain) - probably as bonus subtype, as suggested by @rilian-la-te 
- scouting radius (instant, observatory-like) - if only for support of HotA bank
- executing advmap spell (town gate for example) - might be quite useful, especially if/when advmap spells get more flexible

But not sure if we need:
- random skill/spell/creature _except_ from list/level/school
- movement points _next_ day
Not against them per se, but don't see any real use for such options

**Comment by IvanSavenko** [comment]
Added to list:
- hiding terrain (Cover of Darkness)
- creatures recruitment (dwelling-like)
- limiter for number of empty slots in army

2nd and 3rd are mostly to allow dwelling-like objects, e,g, Ancient Lamp

**Comment by DjWarmonger** [comment]
Spell cast was implemented long time ago: https://github.com/vcmi/vcmi/pull/1928
